File: client/src/Root/RoForge/Agent.lua
-- The agent loop: chat <-> model provider <-> tools, until a final answer.
--
-- This loop runs HERE (in Studio) on purpose: Roblox Studio can only make
-- OUTGOING http requests, so the orchestrator belongs on the client. The model
-- API call is made directly by the provider adapter (Providers/*.lua) — the
-- RoForge backend only executes server-side tools (Tools/RemoteTools.lua).

local Agent = {}

local MAX_TOOL_RESULT_CHARS = 12000

local function truncate(s)
	s = tostring(s or "")
	if #s > MAX_TOOL_RESULT_CHARS then
		return s:sub(1, MAX_TOOL_RESULT_CHARS) .. "\n... [truncated]"
	end
	return s
end

--[[
	cfg:        Config table
	provider:   { chat(cfg, system, history, tools) -> result|nil, err }
	              result = { text: string, calls: { { id, name, args } } }
	history:    (in/out) array of:
	              { role="user", text=string }
	              { role="assistant", text=string?, calls=table? }
	              { role="tool", id=string, name=string, result=string }
	system:     system prompt string
	tools:      array of { name, description, input_schema, run(args) -> string }
	onEvent:    function(kind, payload) with kind in
	              "thinking"|"text"|"tool_start"|"tool_end"|"error"|"done"|"aborted"
	shouldStop: function() -> boolean (user pressed Stop)
]]
function Agent.run(cfg, provider, history, system, tools, onEvent, shouldStop)
	local toolByName = {}
	for _, t in ipairs(tools) do
		toolByName[t.name] = t
	end

	for iter = 1, cfg.MaxIterations do
		if shouldStop and shouldStop() then
			onEvent("aborted")
			return
		end
		onEvent("thinking", iter)

		local result, err = provider.chat(cfg, system, history, tools)
		if err then
			onEvent("error", err)
			return
		end

		local text = result.text or ""
		local calls = result.calls or {}

		if text ~= "" then
			onEvent("text", text)
		end

		if #calls == 0 then
			table.insert(history, { role = "assistant", text = text })
			onEvent("done")
			return
		end

		table.insert(history, { role = "assistant", text = (text ~= "" and text or nil), calls = calls })

		for _, call in ipairs(calls) do
			if shouldStop and shouldStop() then
				onEvent("aborted")
				return
			end
			onEvent("tool_start", call)

			local tool = toolByName[call.name]
			local ok, res
			if not tool then
				ok = false
				res = "ERROR: unknown tool '" .. tostring(call.name) .. "'. Do not call it again."
			else
				ok, res = pcall(tool.run, call.args or {})
				if not ok then
					res = "ERROR: " .. tostring(res)
				end
			end

			local resultStr = truncate(res)
			onEvent("tool_end", { id = call.id, name = call.name, result = resultStr })
			table.insert(history, { role = "tool", id = call.id, name = call.name, result = resultStr })
		end
	end

	onEvent(
		"error",
		("Stopped after %d iterations (safety limit). Ask me to continue where I left off, or narrow the task."):format(
			cfg.MaxIterations
		)
	)
end

return Agent

File: cli/src/agent.js
// The agent loop: chat ↔ model provider ↔ tools, until a final answer.
// Runs locally — streaming, with an iteration safety cap.
import { ProviderError } from "./providers/anthropic.js";

const MAX_TOOL_RESULT_CHARS = 24000;
const MAX_HISTORY_ITEMS = 60;

function truncate(s, n = MAX_TOOL_RESULT_CHARS) {
  s = String(s ?? "");
  return s.length > n ? s.slice(0, n) + `\n... [truncated ${s.length - n} chars]` : s;
}

function trimHistory(history) {
  while (history.length > MAX_HISTORY_ITEMS) history.shift();
  if (history.length && history[0].role !== "user") {
    history.unshift({ role: "user", text: "(Earlier conversation was trimmed to fit the context window.)" });
  }
}

/**
 * run(cfg, provider, history, system, tools, io)
 * provider: { chatStream(cfg, {model, system, messages, tools, signal}, events) -> {text, toolCalls, usage} }
 * io: {
 *   onText(delta), onAssistantDone(text), onToolStart(tool, args), onToolEnd(tool, args, result),
 *   onIter(n, total), onUsage(usage), onDone(), onAborted(), onError(err),
 *   shouldAbort() -> bool,        // user pressed Ctrl+C
 *   approve(name, args) -> bool,  // approval gate (session supplies)
 * }
 * returns { ok, text, usage: {input_tokens, output_tokens}, iterations }
 */
export async function runAgent(cfg, provider, history, system, tools, io) {
  const toolByName = new Map(tools.map((t) => [t.name, t]));
  const maxIterations = cfg.maxIterations || 12;
  let totalUsage = { input_tokens: 0, output_tokens: 0 };
  let finalText = "";

  for (let iter = 1; iter <= maxIterations; iter++) {
    if (io.shouldAbort && io.shouldAbort()) {
      io.onAborted && io.onAborted();
      return { ok: false, aborted: true, text: finalText, usage: totalUsage, iterations: iter - 1 };
    }
    io.onIter && io.onIter(iter, maxIterations);

    let result;
    try {
      result = await provider.chatStream(
        cfg,
        {
          model: cfg._activeModel,
          maxTokens: cfg.maxTokens,
          system,
          messages: history,
          tools,
          signal: io.abortSignal,
        },
        { onText: io.onText }
      );
    } catch (e) {
      if (io.shouldAbort && io.shouldAbort()) {
        io.onAborted && io.onAborted();
        return { ok: false, aborted: true, text: finalText, usage: totalUsage, iterations: iter - 1 };
      }
      io.onError && io.onError(e instanceof ProviderError ? e.message : String(e.message || e));
      return { ok: false, error: true, text: finalText, usage: totalUsage, iterations: iter - 1 };
    }

    if (result.usage) {
      totalUsage.input_tokens += result.usage.input_tokens || 0;
      totalUsage.output_tokens += result.usage.output_tokens || 0;
      io.onUsage && io.onUsage(totalUsage);
    }

    const text = result.text || "";
    const calls = result.toolCalls || [];

    if (text) {
      finalText = text;
    }
    if (!calls.length) {
      history.push({ role: "assistant", text: text || "(no output)" });
      trimHistory(history);
      io.onAssistantDone && io.onAssistantDone(text);
      io.onDone && io.onDone();
      return { ok: true, text: finalText, usage: totalUsage, iterations: iter };
    }

    // record assistant turn with tool calls, then execute each
    history.push({
      role: "assistant",
      text: text || null,
      calls: calls.map((c) => ({ id: c.id, name: c.name, args: c.input })),
    });

    for (const call of calls) {
      if (io.shouldAbort && io.shouldAbort()) {
        io.onAborted && io.onAborted();
        return { ok: false, aborted: true, text: finalText, usage: totalUsage, iterations: iter };
      }
      const tool = toolByName.get(call.name);
      let resultStr;
      let image; // {base64, mediaType} — set when a tool returns an image
      if (!tool) {
        resultStr = `ERROR: unknown tool '${call.name}'. Do not call it again.`;
      } else {
        io.onToolStart && io.onToolStart(tool, call.input || {});
        if (tool.requiresApproval) {
          const ok = (await (io.approve && io.approve(tool.name, call.input || {}))) || false;
          if (!ok) {
            resultStr = "ERROR: user declined to run this tool. Choose a different approach or ask the user what they want.";
            io.onToolEnd && io.onToolEnd(tool, call.input || {}, resultStr);
            history.push({ role: "tool", id: call.id, name: call.name, result: resultStr });
            continue;
          }
        }
        try {
          const raw = await tool.execute(call.input || {}, { cfg, history, iter });
          if (raw && typeof raw === "object") {
            resultStr = String(raw.text ?? "");
            if (raw.image && typeof raw.image.base64 === "string") image = raw.image;
          } else {
            resultStr = String(raw ?? "");
          }
        } catch (e) {
          resultStr = `ERROR: ${e.message || e}`;
        }
      }
      resultStr = truncate(resultStr);
      io.onToolEnd && io.onToolEnd(tool, call.input || {}, resultStr);
      history.push({ role: "tool", id: call.id, name: call.name, result: resultStr, ...(image ? { image } : {}) });
      trimHistory(history);
    }
  }

  const msg = `Stopped after ${maxIterations} iterations (safety limit). Say "continue" to pick up where I left off.`;
  io.onError && io.onError(msg);
  return { ok: false, limit: true, text: finalText, usage: totalUsage, iterations: maxIterations };
}
